garak/translator.py

```diff
@@ -0,0 +1,583 @@
+    """Split input text based on the presence of ': '."""
+    if ": " in input_text:
+        split_text = input_text.splitlines()
+        split_text = [line.split(":") for line in split_text]
+        split_text = [item for sublist in split_text for item in sublist]
+    else:
+        split_text = input_text.splitlines()
+    return split_text
+
+
+def contains_invisible_unicode(text: str) -> bool:
+    """Determine whether the text contains invisible Unicode characters."""
+    if not text:
+        return False
+    for char in text:
+        if unicodedata.category(char) not in {'Cf', 'Cn', 'Zs'}:
+            return False
+    return True
+
+
+def is_nested_list(lst: list) -> bool:
+    """Check if the given list is a nested list."""
+    return any(isinstance(i, list) for i in lst)
+
+
+class SimpleTranslator:
+    """DeepL or NIM translation option"""
+
+    # Reference: https://developers.deepl.com/docs/resources/supported-languages
+    bcp47_deepl = [
+        "ar", "bg", "cs", "da", "de",
+        "en", "el", "es", "et", "fi",
+        "fr", "hu", "id", "it", "ja",
+        "ko", "lt", "lv", "nb", "nl",
+        "pl", "pt", "ro", "ru", "sk",
+        "sl", "sv", "tr", "uk", "zh"
+    ]
+    
+    # Reference: https://docs.nvidia.com/nim/riva/nmt/latest/support-matrix.html#models
+    bcp47_riva = [
+        "zh", "ru", "de", "es", "fr",
+        "da", "el", "fi", "hu", "it",
+        "lt", "lv", "nl", "no", "pl",
+        "pt", "ro", "sk", "sv", "ja",
+        "hi", "ko", "et", "sl", "bg",
+        "uk", "hr", "ar", "vi", "tr",
+        "id", "cs"
+    ]
+    
+    DEEPL_ENV_VAR = "DEEPL_API_KEY"
+    NIM_ENV_VAR = "NIM_API_KEY"
+
+    def __init__(self, plugin_generators_dict: dict={}, source_lang: str="en") -> None:
+        self.translator = None
+        self.nmt_client = None
+        self.translation_service = ""
+        self.source_lang = source_lang
+        self.target_lang = plugin_generators_dict.get("lang_spec", "en")
+        self.translation_service = plugin_generators_dict.get("translation_service", "")
+        self.deepl_api_key = os.getenv(self.DEEPL_ENV_VAR)
+        self.nim_api_key = os.getenv(self.NIM_ENV_VAR)
+
+        if self.translation_service == "deepl" and not self.deepl_api_key:
+            raise ValueError(f"{self.DEEPL_ENV_VAR} environment variable is not set")
+        elif self.translation_service == "nim" and not self.nim_api_key:
+            raise ValueError(f"{self.NIM_ENV_VAR} environment variable is not set")
+
+        self.judge_list = []
+        self._load_translator()
+
+    def _load_translator(self):
+        if self.translation_service == "deepl" and self.translator is None:
+            self.translator = Translator(self.deepl_api_key)
+        elif self.translation_service == "nim" and self.nmt_client is None:
+            auth = riva.client.Auth(None, True, "grpc.nvcf.nvidia.com:443",
+                                    [("function-id", "647147c1-9c23-496c-8304-2e29e7574510"),
```

Comment:
how does the general public get access to this function?

Reply:
`auth` only needs for `riva.client.NeuralMachineTranslationClient`

I think it does not need to general public get access.